File: python/56_MergeIntervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if not intervals: return []
        
        sorted_intervals = sorted(intervals, key = lambda x: x[0])
        curr, merged = sorted_intervals[0], []
        for i in range(1, len(sorted_intervals)):
            if curr[1] >= sorted_intervals[i][0]:
                curr[1] = max(curr[1], sorted_intervals[i][1])
            else:
                merged.append(curr)
                curr = sorted_intervals[i]
        
        merged.append(curr)
        return merged

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if not intervals or not intervals[0]:
            return []
        
        intervals.sort(key = lambda x: x[0])

        res, curr_interval, prev_end, append_end = [], [], None, False
        for [start, end] in intervals:
            if prev_end != None and prev_end >= start:
                curr_interval = [curr_interval[0], max(curr_interval[1], end)]
            else:
                if curr_interval:
                    res.append(curr_interval)
                curr_interval = [start, end]
            
            prev_end = max(end, prev_end) if prev_end else end
            
        return res + [curr_interval]

File: cpp/56_MergeIntervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        if (intervals.empty()) return {};
        
        sort(intervals.begin(), intervals.end(), [](const vector<int>& a, const vector<int>& b) {
            return a[0] == b[0] ? a[1] > b[1] : a[0] < b[0];
        });
            
        vector<vector<int>> merged_intervals;
        vector<int> curr = {intervals[0][0], intervals[0][1]};
        for (int i = 1; i < intervals.size(); ++i) {
            if (curr[1] >= intervals[i][0]) {
                curr[1] = max(curr[1], intervals[i][1]);
            } else {
                merged_intervals.push_back(curr);
                curr = intervals[i];
            }
        }
        merged_intervals.push_back(curr);
        return merged_intervals;
    }
};

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        if (intervals.size() == 0) return {};
        vector<vector<int>> res;
        vector<int> curr_range;
        bool hasStarted = false;
        
        // lambda comparator function to compare each start of interval
        sort(intervals.begin(), intervals.end(), [](const vector<int>& a, const vector<int>& b) {
            return a[0] < b[0];
        });
        for (vector<int> range : intervals) {
            int start = range[0], end = range[1];
            if (!hasStarted) {
                curr_range = {start, end};
                hasStarted = true;
            } else if (curr_range[1] >= start) {
                curr_range[1] = curr_range[1] >= end ? curr_range[1] : end;
            } else {
                res.push_back(curr_range);
                curr_range = {start, end};
            }
        }
        res.push_back(curr_range);
        return res;
    }
};
